src/main/java/vazkii/patchouli/api/data/AbstractPageBuilder.java

```diff
@@ -0,0 +1,48 @@
+package vazkii.patchouli.api.data;
+
+import com.google.gson.JsonObject;
+
+@SuppressWarnings("unchecked")
+public abstract class AbstractPageBuilder<T extends AbstractPageBuilder<T>> {
+	protected final EntryBuilder parent;
+	private final String type;
+	private String advancement;
+	private String flag;
+	private String anchor;
+
+	protected AbstractPageBuilder(String type, EntryBuilder parent) {
+		this.parent = parent;
+		this.type = type;
+	}
+
```

Comment:
- why not just call `super.toJson()` and add the implementation fields there?
- `toJson` can probably be protected too (and would need to be given the above for mod-specific pages)

Reply:
because it's not intended to be called outside of generation and it must be invoked as-is and not depend on super.toJson()

Reply:
I don't agree that's a good reason to have two methods - what does it matter if someone decides to call it outside of generation?

Reply:
the point is it should **ALWAYS** be invoked and if people override it it can't be ensured to still call the super method

Reply:
You're just making it difficult to do things you haven't thought of, while having two similarly named methods in your own code which are only "intended" to run in the same time

Reply:
it's a safety measure that it's never serialized without the required fields

Reply:
But that doesn't prevent it from happening. If you want to ensure it's valid you need to check it in the method that's calling it

Reply:
it doesn't 100% prevent it but it prevents it from happening accidentally and the only way to mess with it is to set the values again in the serialize method so you would have to force it